source/common/router/config_impl.cc

```diff
@@ -720,6 +720,23 @@ void RouteEntryImplBase::finalizeRequestHeaders(Http::RequestHeaderMap& headers,
   }
 }
 
+const VirtualCluster*
+RouteEntryImplBase::virtualCluster(const Http::HeaderMap& headers,
+                                   const StreamInfo::StreamInfo& stream_info) const {
+  const VirtualCluster* virtual_cluster = vhost_.virtualClusterFromEntries(headers);
+  if (virtual_cluster != nullptr && !virtual_cluster->name().has_value()) {
+    // This means virtual cluster has been configured but there is no match in the downstream
+    // headers. Let us check if there is a match in virtual host headers.
+    const auto fake_request_headers = Http::createHeaderMap<Http::RequestHeaderMapImpl>({});
+    for (const HeaderParser* header_parser :
+         getRequestHeaderParsers(vhost_.globalRouteConfig().mostSpecificHeaderMutationsWins())) {
+      header_parser->evaluateHeaders(*fake_request_headers, stream_info);
+    }
+    virtual_cluster = vhost_.virtualClusterFromEntries(*fake_request_headers);
```

Comment:
Not thrilled about having to match the vcluster routes twice for every request just to satisfy a niche use case - should we put this behind a runtime/config flag?

Reply:
I completely agree. That is why in my original implementation moved `finalizeRequestHeaders` to top. I think it is safe but @yanavlasov had some valid concerns. If we go down runtime/config flag - May be we should go back to that solution and add flag. BTW, do you see any issues moving it up or move virtual cluster logic after headers are looked at?

Reply:
Any other better ideas also we can try

Reply:
I think if we change the order of things we need to make sure that makes sense, then runtime guard it in case it has unforeseen consequences, while if we think we the current order is correct then I think we'd want a config flag / disabled by default runtime flag.


@mattklein123 @alyssawilk any thoughts here as router owners?

Reply:
original concerns - https://github.com/envoyproxy/envoy/pull/20648#discussion_r846194754

If we agree, I will put that behind feature flag

Reply:
Another idea. I was after supporting path based metrics. If we can add a `stats_prefix` in `[route match](https://github.com/envoyproxy/envoy/blob/main/api/envoy/config/route/v3/route_components.proto#L457)` - may be we can internally create virtual cluster with that prefix if it exists?

Reply:
I think @yanavlasov concerns are valid, but I would need to look in more detail and do a more complete audit.

With that said, if your goal is per route metrics, this comes up so many times that I would rather just build them in at this point. Obviously this must be opt-in due to metrics explosion issues, but I see no reason to not let people create a stats scope for a route match if they so choose. I think this would be a very popular feature. cc @jmarantz

Reply:
I would prefer to build stats per route.

Reply:
OK let's do that. Do you want to close this and propose an API in a different PR and we can discuss that before implementing?